src/array.cr

```diff
@@ -1545,22 +1544,67 @@ class Array(T)
     self
   end
 
+  # Returns `self` with all the elements shifted `n` times.
+  #
+  # ```
+  # a1 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # a2 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # a3 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  #
+  # a1.rotate!
+  # a2.rotate!(1)
+  # a3.rotate!(3)
+  #
+  # a1 # => [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
+  # a2 # => [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
+  # a3 # => [3, 4, 5, 6, 7, 8, 9, 0, 1, 2]
+  # ```
   def rotate!(n = 1)
     return self if size == 0
+    # This is broken into cases for perfromance (#8515)
     n %= size
-    return self if n == 0
-    if n <= size // 2
+
+    if n == 0
+      # ignore
+    elsif n == 1
+      tmp = self[0]
+      @buffer.move_from(@buffer + n, size - n)
+      self[-1] = tmp
+    elsif n == -1
+      tmp = self[-1]
+      @buffer.move_from(@buffer + n, size - n)
+      self[0] = tmp
+    elsif n <= SMALL_ARRAY_SIZE // 2
+      tmp_buffer = uninitialized StaticArray(T, SMALL_ARRAY_SIZE)
+      tmp_buffer.to_unsafe.copy_from(@buffer, n)
+      @buffer.move_from(@buffer + n, size - n)
+      (@buffer + size - n).copy_from(tmp_buffer.to_unsafe, n)
```

Comment:
`n <= SMALL_ARRAY_SIZE`

Reply:
This should be checking for shift right vs shift left.
I added a comment to be more clear.

Reply:
So if I do `rotate! 9` on the array with size 100, it will not apply the stack-based algorithm, right? It doesn't make any sense. It allocates `SMALL_ARRAY_SIZE` static array, but it is only able to use the half of the array on left rotation.